[MCP 723] Rename `astconv::AstConv` and related items

fmease opened the pull request:
See rust-lang/compiler-team#723.
Corresponding rustc-dev-guide PR: rust-lang/rustc-dev-guide#1916.

Please consult the following *normative* list of changes here:
https://fmease.dev/rustc-dev/astconv-no-mo.html ([2024-03-22 archive link](https://web.archive.org/web/20240322054711/https://fmease.dev/rustc-dev/astconv-no-mo.html)).

fmease committed:
Update (doc) comments

Several (doc) comments were super outdated or didn't provide enough context.

Some doc comments shoved everything in a single paragraph without respecting
the fact that the first paragraph should be a single sentence because rustdoc
treats these as item descriptions / synopses on module pages.

fmease committed:
Update local variables and tracing calls

Most of the tracing calls didn't fully leverage the power of `tracing`.
For example, several of them used to hard-code method names / tracing spans
as well as variable names. Use `#[instrument]` and `?var` / `%var` (etc.) instead.

In my opinion, this is the proper way to migrate them from the old
AstConv nomenclature to the new HIR ty lowering one.

fmease committed:
Rename module astconv to hir_ty_lowering

Split from the main renaming commit to make git generate a proper diff for ease of reviewing.

fmease commented on compiler/rustc_ast_lowering/src/item.rs:
```diff
@@ -1413,8 +1413,8 @@ impl<'hir> LoweringContext<'_, 'hir> {
         // these into hir when we lower thee where clauses), but this makes it quite difficult to
-        // keep track of the Span info. Now, `add_implicitly_sized` in `AstConv` checks both param bounds and
-        // where clauses for `?Sized`.
+        // keep track of the Span info. Now, `<dyn HirLowerer>::add_implicitly_sized` checks both
```
```suggestion
        // keep track of the Span info. Now, `<dyn HirTyLowerer>::add_implicitly_sized` checks both
```

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-16`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/7861277586/job/21449465510) [(plain)](https://github.com/rust-lang/rust/commit/3e1e308054094d7eee84cd293903164b305ec077/checks/21449465510/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#12 writing image sha256:b8775847f59473b086583855ad5930603613e952eaac21d11a3a65b6a63c8e1e done
#12 naming to docker.io/library/rust-ci done
#12 DONE 10.1s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-16]
##[group]Clock drift check
  local time: Sun Feb 11 10:18:35 UTC 2024
  network time: Sun, 11 Feb 2024 10:18:35 GMT
  network time: Sun, 11 Feb 2024 10:18:35 GMT
##[endgroup]
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-16', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'build.optimized-compiler-builtins', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-16/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
   Compiling rustc_hir_analysis v0.0.0 (/checkout/compiler/rustc_hir_analysis)
error[E0433]: failed to resolve: use of undeclared crate or module `trace`
  --> compiler/rustc_hir_analysis/src/lowering/object_safety.rs:30:22
   |
30 |         let _guard = trace::debug_span!("lower_object_ty_to_poly_trait_ref");
   |                      ^^^^^ use of undeclared crate or module `trace`
For more information about this error, try `rustc --explain E0433`.
error: could not compile `rustc_hir_analysis` (lib) due to 1 previous error
warning: build failed, waiting for other jobs to finish...
Build completed unsuccessfully in 0:01:56
```

</details>

fmease commented on compiler/rustc_hir_analysis/src/astconv/mod.rs:
```diff
@@ -1149,7 +1140,6 @@ impl<'o, 'tcx> dyn HirTyLowerer<'tcx> + 'o {
     /// parameter or `Self`.
     // NOTE: When this function starts resolving `Trait::AssocTy` successfully
     // it should also start reporting the `BARE_TRAIT_OBJECTS` lint.
-    #[instrument(level = "debug", skip(self, hir_ref_id, span, qself, assoc_segment), fields(assoc_ident=?assoc_segment.ident), ret)]
```
I think this needs to be reverted since `#[instrument]` plays nicer with `parallel_rustc` afaiu.

fmease commented on compiler/rustc_hir_analysis/src/lib.rs:
```diff
@@ -214,18 +214,23 @@ pub fn check_crate(tcx: TyCtxt<'_>) -> Result<(), ErrorGuaranteed> {
-/// A quasi-deprecated helper used in rustdoc and clippy to get
-/// the type from a HIR node.
-pub fn hir_ty_to_ty<'tcx>(tcx: TyCtxt<'tcx>, hir_ty: &hir::Ty<'tcx>) -> Ty<'tcx> {
+/// Helper to lower a [`hir::Ty`] to a [`Ty`].
```
Add back the mention of rustdoc and Clippy. I originally removed it since there were some uses in `rustc_privacy` until very recently. Now the note makes sense again.

fmease reviewed (commented):
Also I'm relatively sure that `instantiate_{mono,poly}_trait_ref` should actually be called `lower_$morph_trait_ref` since it's not really instantiating anything (with generic args). NB: There exists `lower_path_to_mono_trait_ref`. Check how they differ.

bors commented:
:umbrella: The latest upstream changes (presumably #120951) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

fmease commented on compiler/rustc_hir_analysis/src/astconv/bounds.rs:
```diff
@@ -234,11 +234,11 @@ impl<'tcx> dyn AstConv<'tcx> + '_ {
     /// the binder (e.g., `&'a u32`) and hence may reference bound regions.
     #[instrument(level = "debug", skip(self, bounds, speculative, dup_bindings, path_span))]
-    pub(super) fn add_predicates_for_ast_type_binding(
+    pub(super) fn lower_type_bindings_to_predicates(
```
```suggestion
    pub(super) fn lower_type_binding_to_predicates(
```

or even:

```suggestion
    pub(super) fn lower_assoc_item_binding_to_predicates(
```

bors commented:
:umbrella: The latest upstream changes (presumably #121252) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/7955599992/job/21714828261) [(plain)](https://github.com/rust-lang/rust/commit/3afdd34b04eb317fee7c7b343d89a5bedeca80c2/checks/21714828261/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#13 3.971 Building wheels for collected packages: reuse
#13 3.972   Building wheel for reuse (pyproject.toml): started
#13 4.307   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 4.308   Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=181117 sha256=f5f58750481f69515c2c0d1d503daf565e2565c370d07fc6aeb95fe3498b4269
#13 4.308   Stored in directory: /tmp/pip-ephem-wheel-cache-5etr73i6/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
#13 4.311 Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
#13 4.334   Attempting uninstall: setuptools
#13 4.335     Found existing installation: setuptools 59.6.0
#13 4.336     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
---
 Documenting rustc_lint v0.0.0 (/checkout/compiler/rustc_lint)
error: could not parse code block as Rust code
   --> compiler/rustc_hir_analysis/src/hir_ty_lowering/bounds.rs:102:5
    |
102 | /     /// Lower HIR bounds into `bounds` given the self type `param_ty`.
103 | |     ///
104 | |     /// ### Example
...   |
...   |
115 | |     /// There is an implied binder around `param_ty` and `hir_bounds`.
116 | |     /// See `lower_poly_trait_ref` for more details.
    |
    |
    = help: `ignore` code blocks require valid Rust code for syntax highlighting; mark blocks that do not contain Rust code as text: ```text
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: error from rustc: unknown start of token: `
    = note: `-D rustdoc::invalid-rust-codeblocks` implied by `-D warnings`
    = note: `-D rustdoc::invalid-rust-codeblocks` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(rustdoc::invalid_rust_codeblocks)]`

error: could not document `rustc_hir_analysis`

Caused by:
  process didn't exit successfully: `/checkout/obj/build/bootstrap/debug/rustdoc --edition=2021 --crate-type lib --crate-name rustc_hir_analysis compiler/rustc_hir_analysis/src/lib.rs --target x86_64-unknown-linux-gnu -o /checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/doc -Zunstable-options --check-cfg 'cfg(feature, values())' --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat -C metadata=600935aa135de830 -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/release/deps --extern itertools=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libitertools-57254199b2c46c56.rmeta --extern rustc_arena=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_arena-4482f014cf87dff6.rmeta --extern rustc_ast=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_ast-5e0e71d08d9290fd.rmeta --extern rustc_attr=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_attr-d46d3bac488a40c3.rmeta --extern rustc_data_structures=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_data_structures-d31de55e71c39494.rmeta --extern rustc_errors=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_errors-801d365d8d2176c3.rmeta --extern rustc_feature=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_feature-5d2c8ac6f0d7f305.rmeta --extern rustc_fluent_macro=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/release/deps/librustc_fluent_macro-fe3541ce532d5a46.so --extern rustc_hir=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_hir-a5abf01e2a45d6ea.rmeta --extern rustc_hir_pretty=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_hir_pretty-b2fbcaaa37e82885.rmeta --extern rustc_index=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_index-9072a0c9c38984bd.rmeta --extern rustc_infer=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_infer-f7ee5e8fb6902bf8.rmeta --extern rustc_lint_defs=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_lint_defs-ff6744b770ddc98a.rmeta --extern rustc_macros=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/release/deps/librustc_macros-657bde1f4a8bb06b.so --extern rustc_middle=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_middle-68758be9e96ce72d.rmeta --extern rustc_session=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_session-2fbce4160c832450.rmeta --extern rustc_span=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_span-38d5fe1c5a028c70.rmeta --extern rustc_target=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_target-2a813ec15f2f25cf.rmeta --extern rustc_trait_selection=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_trait_selection-58060336a9b2a7ea.rmeta --extern rustc_type_ir=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_type_ir-384634a5dedf106c.rmeta --extern smallvec=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libsmallvec-17d63ea25df2dc47.rmeta --extern tracing=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libtracing-2cfab5196b215b73.rmeta --extern-html-root-url 'itertools=https://docs.rs/itertools/0.11.0/' --extern-html-root-url 'smallvec=https://docs.rs/smallvec/1.13.1/' --extern-html-root-url 'tracing=https://docs.rs/tracing/0.1.37/' -Zunstable-options --cfg=bootstrap --cfg=windows_raw_dylib -Csymbol-mangling-version=v0 -Zunstable-options '--check-cfg=cfg(bootstrap)' '--check-cfg=cfg(parallel_compiler)' '--check-cfg=cfg(no_btreemap_remove_entry)' '--check-cfg=cfg(crossbeam_loom)' '--check-cfg=cfg(span_locations)' '--check-cfg=cfg(rustix_use_libc)' '--check-cfg=cfg(emulate_second_only_system)' '--check-cfg=cfg(windows_raw_dylib)' --document-private-items --document-hidden-items -Dwarnings '-Wrustdoc::invalid_codeblock_attributes' --crate-version '1.78.0-nightly (ed6cfe173 2024-02-19)' --document-private-items '-Arustdoc::private-intra-doc-links' --enable-index-page -Zunstable-options -Znormalize-docs --show-type-layout --generate-link-to-definition '-Zcrate-attr=warn(rust_2018_idioms)' --cfg=parallel_compiler --extern-html-root-url 'ena=https://docs.rs/ena/latest/'` (exit status: 1)
Build completed unsuccessfully in 0:01:05
  local time: Mon Feb 19 06:57:43 UTC 2024
  network time: Mon, 19 Feb 2024 06:57:43 GMT
##[error]Process completed with exit code 1.
```

</details>

fmease commented on compiler/rustc_hir_analysis/src/astconv/bounds.rs:
```diff
@@ -98,21 +99,21 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
+    /// fn foo<T: Debug>() { }
+    ///        ^  ^^^^^ `ast_bounds`, in HIR form
     ///        |
     ///        `param_ty`, in ty form
```
```suggestion
    /// //     ^  ^^^^^ `ast_bounds`, in HIR form
    /// //     |
    /// //     `param_ty`, in ty form
```

bors commented:
:umbrella: The latest upstream changes (presumably #120463) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

fmease commented on compiler/rustc_hir_analysis/src/lib.rs:
```diff
@@ -231,36 +228,10 @@ pub fn check_crate(tcx: TyCtxt<'_>) -> Result<(), ErrorGuaranteed> {
-    item_cx.astconv().ast_ty_to_ty(hir_ty)
-}
-
-pub fn hir_trait_to_predicates<'tcx>(
```
Unused since #113671.

bors commented:
:umbrella: The latest upstream changes (presumably #121489) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #122139) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

fmease committed:
Notify fmease on changes to HIR ty lowering

fmease commented:
> be398be414cb3479feb5fb04e1117d39aab2e5d2

Might've affected perf (I doubt it tho). Better safe that sorry, let's check it.
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 1fa7d956130d75b653d8e97d1899275945108998 with merge df71c4f955d124740ac58ef188d8c9bd24812f20...
<!-- homu: {"type":"TryBuildStarted","head_sha":"1fa7d956130d75b653d8e97d1899275945108998","merge_sha":"df71c4f955d124740ac58ef188d8c9bd24812f20"} -->

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

compiler-errors commented:
Regardless, be398be414cb3479feb5fb04e1117d39aab2e5d2 doesn't seem just "rename". Why not pull anything that changes signatures into a separate PR?

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/8284127293/job/22672798179)
Build commit: df71c4f955d124740ac58ef188d8c9bd24812f20 (`df71c4f955d124740ac58ef188d8c9bd24812f20`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/8284127293/job/22672798179"},"merge_sha":"df71c4f955d124740ac58ef188d8c9bd24812f20"} -->

rust-timer commented:
Queued df71c4f955d124740ac58ef188d8c9bd24812f20 with parent c2fbe404d21572a04dabc6a0e23eb3ab367832ad, future [comparison URL](https://perf.rust-lang.org/compare.html?start=c2fbe404d21572a04dabc6a0e23eb3ab367832ad&end=df71c4f955d124740ac58ef188d8c9bd24812f20).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.5 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (df71c4f955d124740ac58ef188d8c9bd24812f20): [comparison URL](https://perf.rust-lang.org/compare.html?start=c2fbe404d21572a04dabc6a0e23eb3ab367832ad&end=df71c4f955d124740ac58ef188d8c9bd24812f20&stat=instructions:u).

### Overall result: no relevant changes - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=c2fbe404d21572a04dabc6a0e23eb3ab367832ad&end=df71c4f955d124740ac58ef188d8c9bd24812f20&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=c2fbe404d21572a04dabc6a0e23eb3ab367832ad&end=df71c4f955d124740ac58ef188d8c9bd24812f20&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=c2fbe404d21572a04dabc6a0e23eb3ab367832ad&end=df71c4f955d124740ac58ef188d8c9bd24812f20&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=c2fbe404d21572a04dabc6a0e23eb3ab367832ad&end=df71c4f955d124740ac58ef188d8c9bd24812f20&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 670.541s -> 670.004s (-0.08%)
**Artifact size**: 310.79 MiB -> 310.78 MiB (-0.00%)

fmease committed:
Rename AstConv to HIR ty lowering

This includes updating astconv-related items and a few local variables.

bors commented:
:umbrella: The latest upstream changes (presumably #122607) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #121123) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

fmease committed:
Arrange methods on HirTyLowerer more logically

This makes it easier to read the trait definition for newcomers:
Sorted from least “complex” to most “complex” followed by trivial “plumbing”
and grouped by area.

* Move `allow_infer` above all `*_infer` methods
  * It's the least complex method of those
  * Allows the `*_infer` to be placed right next to each other
* Move `probe_ty_param_bounds` further down right next to `lower_assoc_ty` and `probe_adt`
  * It's more complex than the `infer` methods, it should come “later”
  * Now all required lowering functions are grouped together
* Move the “plumbing” function `set_tainted_by_errors` further down
  below any actual lowering methods.
* Provided method should come last

fmease commented:
Alrighty, the renaming and updating of all relevant (doc) comments has been completed at last!

Best reviewed commit by commit. Some commits contain a short explainer outlining the motivations for the respective changes if not entirely obvious.

Please consult the following *normative* list of changes here: https://fmease.dev/rustc-dev/astconv-no-mo.html ([2024-03-22 archive link](https://web.archive.org/web/20240322054711/https://fmease.dev/rustc-dev/astconv-no-mo.html)). The [MCP](https://github.com/rust-lang/compiler-team/issues/723) doesn't contain every single change but only the majority of changes.

Feel free to suggest better names for individual items but keep in mind that most of the name changes are well motivated and are part of a cohesive naming scheme. If you do please double-check first if there's an explainer for the name change in question over at the linked web page.

Ping @oli-obk @compiler-errors

oli-obk commented:
<3 Thanks for doing this. The html doc is amazing. Still reading the doc, then will give this PR a review

@bors p=10 bitrotty

oli-obk commented on compiler/rustc_trait_selection/src/traits/object_safety.rs:
```diff
@@ -37,6 +37,7 @@ pub use crate::traits::{MethodViolationCode, ObjectSafetyViolation};
 /// Currently that is `Self` in supertraits. This is needed
 /// because `object_safety_violations` can't be used during
 /// type collection.
+#[instrument(level = "debug", skip(tcx))]
```
could also use `ret` here

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit 5e73a8b2f149a359abb2717d5fa91f63add75eef has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk 5e73a8b2f149a359abb2717d5fa91f63add75eef -->
<!-- homu: {"type":"Approved","sha":"5e73a8b2f149a359abb2717d5fa91f63add75eef","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

fmease commented on compiler/rustc_hir_analysis/src/astconv/bounds.rs:
```diff
@@ -101,21 +103,27 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
-    /// **A note on binders:** there is an implied binder around
-    /// `param_ty` and `ast_bounds`. See `instantiate_poly_trait_ref`
-    /// for more details.
+    /// There is an implied binder around `param_ty` and `ast_bounds`.
```
It just means that `param_ty: Ty<'tcx>` may contain late-bound vars even though it's not wrapped in a `Binder<'tcx, ·>`. Neither is `ast_bounds`/`hir_bounds` but well that param is of a HIR data type, so obviously it can't be wrapped in a `Binder`.

Wording is a bit poor, I guess I could've adjusted it.

bors commented:
:hourglass: Testing commit 5e73a8b2f149a359abb2717d5fa91f63add75eef with merge eff958c59e8c07ba0515e164b825c9001b242294...
<!-- homu: {"type":"BuildStarted","head_sha":"5e73a8b2f149a359abb2717d5fa91f63add75eef","merge_sha":"eff958c59e8c07ba0515e164b825c9001b242294"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/8388916172/job/22978165935)
Approved by: oli-obk
Pushing eff958c59e8c07ba0515e164b825c9001b242294 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"oli-obk","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/8388916172/job/22978165935"},"merge_sha":"eff958c59e8c07ba0515e164b825c9001b242294"} -->

rust-timer commented:
Finished benchmarking commit (eff958c59e8c07ba0515e164b825c9001b242294): [comparison URL](https://perf.rust-lang.org/compare.html?start=eb80be223f39d083df649fa3eaff428bfbad9599&end=eff958c59e8c07ba0515e164b825c9001b242294&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=eb80be223f39d083df649fa3eaff428bfbad9599&end=eff958c59e8c07ba0515e164b825c9001b242294&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=eb80be223f39d083df649fa3eaff428bfbad9599&end=eff958c59e8c07ba0515e164b825c9001b242294&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=eb80be223f39d083df649fa3eaff428bfbad9599&end=eff958c59e8c07ba0515e164b825c9001b242294&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 2.2% | [2.2%, 2.2%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=eb80be223f39d083df649fa3eaff428bfbad9599&end=eff958c59e8c07ba0515e164b825c9001b242294&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 669.305s -> 669.088s (-0.03%)
**Artifact size**: 315.01 MiB -> 314.99 MiB (-0.01%)
